Fix #3505: Implement crop on filepath editor

juansaba opened the pull request:
This is the last step to implement the feature described in issue #3505.

juansaba committed:
Implemented image cropping in the filepath editor

juansaba committed:
Fixed lint error

juansaba committed:
Fixed lint errors

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -34,36 +34,125 @@ oppia.directive('filepathEditor', [
+
+        /** Internal functions (not shared with the view) */
+
+        var getResampledImageData = function(imageData, width, height) {
```
Could you document the args here?

juansaba replied:
Done.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -34,36 +34,125 @@ oppia.directive('filepathEditor', [
+          return canvas.toDataURL('image/' + OUTPUT_IMAGE_FORMAT, 1);
+        };
+
+        var getCroppedImageData = function(imageData, x, y, width, height) {
```
Ditto (re documentation). Otherwise it's a bit unclear what exactly width/height are referring to and what the format of imageData is.

juansaba replied:
Done.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -34,36 +34,125 @@ oppia.directive('filepathEditor', [
         };
-        $scope.clearScopeData();
+
+        var getCorrectedMouseEventCoordinates = function(e) {
```
"Corrected" doesn't really make sense without context. Perhaps getEventCoordsRelativeToImageContainer()?

juansaba replied:
Done.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -73,27 +162,321 @@ oppia.directive('filepathEditor', [
+          return cropWidth < dimensions.width || cropHeight < dimensions.height;
+        };
+
+        $scope.mouseMoveWhileDraggingCroppingArea = function(x, y) {
```
This is an internal handler (no need to share with view), so use var.

It should also start with a verb: handleMouseMoveWhileDraggingCroppingArea() or handleMouseMoveWhileDraggingCropArea().

juansaba replied:
Done.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -73,27 +162,321 @@ oppia.directive('filepathEditor', [
+          $scope.croppingArea.y2 = y1 + cropHeight;
+        };
+
+        $scope.mouseMoveWhileResizingCroppingArea = function(x, y) {
```
Ditto (naming and var).

juansaba replied:
Done.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -73,27 +162,321 @@ oppia.directive('filepathEditor', [
+        $scope.mouseMoveWhileDraggingCroppingArea = function(x, y) {
+          var xDown = $scope.mouseLastDownCoordinates.x;
+          var yDown = $scope.mouseLastDownCoordinates.y;
+          var x1WhenDown = $scope.croppingAreaXWhenLastDown;
```
"When last down" feels a bit ambiguous. Maybe "...WhenMouseLastDown"?

Incidentally, I'm mildly starting to wonder whether we should introduce a Point domain object (with 'x' and 'y' attrs), but up to you.

juansaba replied:
I renamed it to "lastMouseDownEventCoordinates". The naming was from the beginning referring to the MouseDown event, so that's why I used it. Let me know what you think.

I see what you're saying about Point. Maybe if it was going to be used globally across Oppia, but if it is for just this component, I'd find it overkill.

seanlip replied:
Oh, that makes sense. I like the name you chose!

And OK, happy to defer to you re Point.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -73,27 +162,321 @@ oppia.directive('filepathEditor', [
+
+          var adjustResizeLeft = function(x) {
+            // Update crop area x1 value, correcting for boundaries.
+            $scope.croppingArea.x1 = Math.min(
```
Looks like you use this construct a lot -- I suggest defining a helper method `clamp(value, minBoundary, maxBoundary)` and calling it via `$scope.croppingArea.x1 = clamp(x, 0, $scope.croppingArea.x2 - CROP_AREA_MIN_WIDTH_PX);`.

juansaba replied:
Done.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -73,27 +162,321 @@ oppia.directive('filepathEditor', [
+          } else if (xOnLeftBorder && yOnBottomBorder) {
+            // Lower left corner.
+            $scope.mousePositionWithinCroppingArea = MOUSE_BOTTOM_LEFT;
+          } else if (xOnLeftBorder) {
```
Style: perhaps might be cleaner to do all the corner checks, followed by all the edge checks, followed by the inside check.

juansaba replied:
Well, I did it in the same order as you define CSS borders, i.e. clock-wise. It's not random.

If you feel strongly about it I can change it.

seanlip replied:
Hm, are you sure? I don't think CSS border order is "top-left then bottom-left then left then top-right then bottom-right..."

The main reason I wanted to double-check this is that the checks overlap, so the order matters. I figured you'd want to do the stricter corner checks before the less strict edge checks. But if you prefer the current ordering, I don't feel strongly.

juansaba replied:
No, you're right. And I looked at the wrong code section. I do clock-wise somewhere else, but here it's kind of random and you SHOULD check the corners first, otherwise we're making the corners smaller than they should be and harder to grab.

Done!

I've added an implementation comment as well.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -73,27 +162,321 @@ oppia.directive('filepathEditor', [
+          $scope.data.metadata.originalWidth = width;
+          $scope.data.metadata.originalHeight = height;
+
+          // Re-calculate dimensions reset cropping area to the boundaries.
```
Is there a missing "and" in this sentence?

Also, slight rewording suggestion: "... and reset the boundaries of the cropping area."

juansaba replied:
Well the point was to set the coordinates of the crop area "to the boundaries" of the image. I have rephrased this. Let me know if it's any better.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -73,27 +162,321 @@ oppia.directive('filepathEditor', [
+          }
+        };
+
+        $scope.mouseMoveOnImageArea = function(e) {
```
Event handler names should start with on -- e.g. onMouseMoveOnImageArea(), or just onMouseMove() in this case.

juansaba replied:
Done.

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -73,27 +162,321 @@ oppia.directive('filepathEditor', [
+          $scope.mouseLastKnownCoordinates = {x: coords.x, y: coords.y};
+        };
+
+        $scope.mouseDownOnCroppingArea = function(e) {
```
onCropAreaMousedown() or onCroppingAreaMousedown() or onMousedownInCroppingArea() perhaps.

(Just FYI, "crop area" instead of "cropping area" sounds fine to me and is shorter, but I don't really have a preference.)

juansaba replied:
Renamed all cropping to crop. Except some obvious one, like "isUserCropping" :)

seanlip commented on extensions/objects/templates/FilepathEditor.js:
```diff
@@ -73,27 +162,321 @@ oppia.directive('filepathEditor', [
+          }
+        };
+
+        $scope.mouseUpOnCroppingArea = function(e) {
```
Ditto: naming.

juansaba replied:
Done.

seanlip reviewed (changes requested):
Thanks, @juansaba! I've gone through the code and left some comments. I'll add some UI comments to the main thread.

seanlip commented:
UI comments: Overall, it looks nice and behaves well! I think I found a few bugs though.

(a) When testing in both Firefox and Chrome (on Linux Ubuntu), I noticed that if I upload an image and drag the top-right border around, sometimes the crop area snaps randomly to a small area on the left side of the screen. In the following screenshot, the location of my mouse is actually somewhere on the right of the image, near the word "from". The width of the "cropped" area is roughly always the same when this happens (in case this helps with debugging).

![crop-area-in-wrong-place](https://user-images.githubusercontent.com/10575562/27775087-2b0217ca-5f53-11e7-88fb-98f0df59fe1e.png)

(b) There is a bit of rapid juddering when "Use this image" is clicked, and when an image is discarded. This is a bit disconcerting.

(c) I uploaded [this picture](https://user-images.githubusercontent.com/10575562/27775136-d5f84942-5f53-11e7-8c09-1e2879b98b5a.png) and was playing around with it in Firefox -- note that it has a transparent background. I did a crop-and-save, and it ended up like this (with a red background). I had difficulty reproducing this reliably, though, but I remember one time later, after crop, I saw that a thin red line had been introduced on the base of the cropped image (I didn't take a screenshot then though).
![oops](https://user-images.githubusercontent.com/10575562/27775142-e87a4520-5f53-11e7-90f3-f1a87e27c2af.png)

(d) In a fresh exploration, open the RTE, upload an image, add a caption, and save. Then upload another image, resize the crop area (but don't click Crop/Cancel), and then Cancel the full image modal (without clicking "use this image"). Then save the content. The first image also disappears (although it shouldn't). This only seems to happen in Firefox.

juansaba committed:
Addressing review comments

juansaba committed:
Fix lint issue

juansaba commented:
@seanlip 

(a) I tried to reproduce this on Chrome and Firefox. Couldn't do it. I used Mac OS X though. I don't have a Ubuntu environment to test :(

(b) Do you mean the transition from showing the uploaded image to showing the saved image? I believe this is because of the time it takes to load the saved image. In order to mitigate this, I've made it so that the final image URL is "loaded" in memory before the mode is changed to MODE_SAVED. This way, when the view changes, the image is already loaded. I think it did improve. Let me know what you think.

(c) Nice catch! I can investigate this a bit more and figure out how to deal with transparency. But before I do that, can you think of a scenario where replacing transparency with a white background wouldn't work? I've currently implemented a really simple one-line fix, where transparent images will end up having a white background. Let me know what you think.

EDIT: thinking about this more, if we reuse this later on for the preferences, we will want transparency to become transparent, not white. Cause the toolbar at the top is not white. It's green. So feel free to push back on this, or we can implement the transparency thing "if and when" we migrate the preferences loader. Wdyt?

(d) Again I followed your steps but couldn't reproduce the results :(

seanlip commented:
Hi @AllanYangZhou -- would you mind taking a look at the UI for this? Any comments you have on the UX would be welcome. Also, would you mind checking out the bugs I mentioned above but that @juansaba is not able to reproduce (if you have a Linux machine)?

Thanks!

seanlip commented:
For (c), let's make sure we deal with transparency correctly, and use a transparent background.

juansaba committed:
Addressed additional comment

juansaba commented:
@seanlip I have addressed the comment about checking pointer position on borders. Thanks!

As discussed, I will work on correctly handling transparent background on the output image.

Stay tuned!

seanlip commented:
Btw, I tested this with the latest changes (Firefox on Linux).

Issue (a) still occurs for me, quite reliably.

I can't test issue (d), because there's a new problem for me. After uploading an image and filling in all the fields, I can't save the image. The "Done" button remains disabled.

juansaba committed:
Fixed background of transparent images

juansaba commented:
@seanlip I noticed the Save image problem. My bad. I just fixed it.

Also I fixed the red background issue. Turns out that the resampled image was fine. It had transparent background all along, but the <img> tag that hosts the image after it is saved had red background don't know why. So I just set the background of that <img> tag to white, which works because this is only used on the filepath editor. Afterwards, when you actually save the image, it is loaded as transparent. You can open it in a new tab and see that it is actually transparent.

So, right now what's pending is:
* Double check (a)
* For you to double check (d)

Please be specific on the steps to reproduce if you confirm both these issues, so that I can follow them exactly like you did.

seanlip commented:
(a) still happens reliably on my Ubuntu machine. To reproduce:
1. Sign up as a new user in Firefox and create a new exploration.
2. Click on the content field that shows up and add an image.
3. Click "Upload a File" and select an image from the computer. (I used [this one](https://user-images.githubusercontent.com/10575562/27812347-15a9ee5e-6023-11e7-8ea5-b627721148ad.png).)
4. Grab the top-right corner of the cropping rectangle with your mouse and move it around.

Observed: The cropping rectangle flickers while dragging, and sometimes becomes a narrow strip on the left of the image.
Expected: The top-right of the cropping rectangle stays where the mouse pointer is.

seanlip commented:
(d) also still happens reliably on my Ubuntu machine. To reproduce:

1. Sign up as a new user in Firefox and create a new exploration.
2. Click on the content field that shows up and add an image.
3. Click "Upload a File" and select an image from the computer.
4. Click "Use This Image", add a "visually impaired" explanation and click "Done".
5. Click the "Add Image" button.
6. Click "Upload a File" and select a different image from the computer.
7. Grab the top-right corner of the cropping rectangle with your mouse and move it around. **Do not** click "Crop" or "Cancel".
8. Click "Cancel" at the very bottom of the modal. The (editable) rich-text editor now shows the first image.
9. Click the green "Save Content" button.

Observed: The first image disappears.
Expected: The first image should remain in the RTE.

juansaba commented:
@seanlip 

I was able to reproduce (a) and I'm working on fixing it.

I was able to reproduce (d) as well, but it also occurs at www.oppia.org right now, which doesn't even have the new image uploader changes. So I believe this is a bug on the RTE or the way the RTE is being used.

You can go to www.oppia.org and do the following:

1. Create an exploration
2. Open the RTE and click the "add an image" button
3. Select an image from your computer and press "Save", add a vision impaired text and press "Done".
4. Click the "add an image" button again, click on "select a file" and select a file from your computer.
5. Without pressing "Save", go and click the "Cancel" button at the bottom of the modal.
6. Click on "Save Content"

The first image will have disappeared.

seanlip commented:
Thanks! Filed #3604.

juansaba committed:
Fixed identification of (x, y) coordinates associated to mousemove events

juansaba commented:
@seanlip I just pushed the fix for (a).

This was due to the function that identifies the correct mouse event coordinates being flaky. I've re-written this function in a much more robust way.

seanlip commented:
Thanks, @juansaba -- this looks great! I only see one issue. When clicking "Use this image", there is a violent judder: the image momentarily flashes below its original position, in the region of the caption/alt-text fields. Can this be fixed?

Other than that, LGTM :)

juansaba committed:
Changed ng-show to ng-if for consistency

seanlip commented:
Discussed the flickering issue with @juansaba. Looks like an Angular bug -- see https://github.com/angular/angular.js/issues/14015#issuecomment-256604563 which reports a similar issue. Leaving as-is for now, so this PR LGTM.

AllanYangZhou reviewed (approved):
Just played around with this, looks really good! I can merge after the automated checks.
